### ISSUE TITLE
There should be matchers for std::optional<T>

### ISSUE THREAD

**Issue by akonradi** [google/googletest#1644]
It would be nice to have a `HasValue(m)` matcher for asserting that an optional object has a value that matches the given inner matcher `m`. It would also be nice to have a matcher for asserting that an optional was empty, maybe `HasNoValue` or `IsNullOpt`, though it could also be `HasValue()` without any arguments.

I'd be happy to contribute this if it sounds useful.

**Comment by sbenzaquen** [comment]
On Tue, Jun 26, 2018 at 1:58 PM Alex Konradi <notifications@github.com>
wrote:

> It would be nice to have a HasValue(m) matcher for asserting that an
> optional object has a value that matches the given inner matcher m.
>
This is called Optional(m)

> It would also be nice to have a matcher for asserting that an optional was
> empty, maybe HasNoValue or IsNullOpt, though it could also be HasValue()
> without any arguments.
>
You can do this with Eq(nullopt)

There is also a matcher for variant, called VariantWith. Eg
VariantWith<int>(Ge(7))

**Comment by akonradi** [comment]
Oops, thanks for pointing out `Optional(m)`. I couldn't find it anywhere in the "For Dummies" or "Cookbook" docs, so that might be helpful to add.

**Comment by adambadura** [comment]
> This is called Optional(m)

Where is this `Optional(m)` defined? Documentation doesn't mention it (as already pointed out by @akonradi). However, I also fail to find it in source code...

> You can do this with Eq(nullopt)

Sure, this is possible. Much like `Eq(nullptr)` is possible (or could be I guess). And yet we still have `IsNull` and `NotNull`. Having `IsNone` or `NotNone` in a similar way for `optional` could be beneficial for better descriptiveness and cooperation with various `optional` implementations (including obviously the one from Boost).

**Comment by sbenzaquen** [comment]
On Fri, Jul 6, 2018 at 6:36 AM Adam Badura <notifications@github.com> wrote:

> This is called Optional(m)
>
> Where is this Optional(m) defined? Documentation doesn't mention it (as
> already pointed out by @akonradi <https://github.com/akonradi>). However,
> I also fail to find it in source code...
>
You are right. For some reason is missing in the documentation. It should
be in
https://github.com/google/googletest/blob/master/googlemock/docs/CheatSheet.md#generic-comparison
right before `VariantWith`.

> You can do this with Eq(nullopt)
>
> Sure, this is possible. Much like Eq(nullptr) is possible (or could be I
> guess). And yet we still have IsNull and NotNull.
>
We have IsNull and NotNull because these were added before nullptr was a
thing.
You could not do Eq(NULL) as type deduction would have chosen the wrong
type there and the comparison would have failed.
`Eq(nullptr)` is an acceptable way of doing it today. Also note that `Eq`
is implicit for the most part so you could just say `nullptr` and that will
be a equivalent to `Eq(nullptr)`.

> Having IsNone or NotNone in a similar way for optional could be
> beneficial for better descriptiveness and cooperation with various
> optional implementations (including obviously the one from Boost).
>
Having less matchers is good too because there is less to learn.
You already know what ==nullptr / !=nullptr mean, so Eq(nullptr) and
Ne(nullptr) come for free without having to learn `IsNull` and `NotNull`.

**Comment by akonradi** [comment]
I was using the Optional matcher again today and realized I never circled back here. I'll open a pull request to add it to the documentation.

**Comment by gennadiycivil** [comment]
@akonradi ping

**Comment by adambadura** [comment]
@akonradi, what about matcher verifying whether `optional` has no value (something like `IsNull` or corresponding `NotNull`)? Do we have any? Or will the pointer ones work?

Expected benefit of such matcher (same as `Optional`) is that it will work with various kinds of `optional` types (`std::optional`, `std::experimental::optional`, `boost::optional` - to name a few) unlike primitive `Eq` which will not.

**Comment by akonradi** [comment]
@adambadura I've been using `Eq(absl::nullopt)` with `absl::optional` and it's been working just fine. This works for me but it does require using `nullopt` of the correct type.

**Comment by adambadura** [comment]
@akonradi, yes, this is what I meant. `Eq` or `Ne` are fine however, they bind to a specific implementation. While this is not a big issue it is still something worth (IMHO) taking care of.

**Comment by sbenzaquen** [comment]
On Tue, Oct 30, 2018 at 10:31 AM Adam Badura <notifications@github.com>
wrote:

> @akonradi <https://github.com/akonradi>, yes, this is what I meant. Eq or
> Ne are fine however, they bind to a specific implementation. While this
> is not a big issue it is still something worth (IMHO) taking care of.
>
We could have a matcher that triggers the explicit conversion to bool.
That would work with all versions of optional, and would also work with
other types explicitly convertible to bool like smart pointers and
std::function.
Not sure if the generic matcher is actually worth it, though. It is more
stuff to remember and would only really matter in generic code.
